internal/config/rule.go

```diff
@@ -1,195 +1,124 @@
-							recipient.Recipient = g
-						} else {
-							recipientLogger.Warnw("ignoring unknown escalation recipient")
-							escalation.Recipients[i] = nil
-						}
-					} else if recipient.ScheduleID.Valid {
-						id := recipient.ScheduleID.Int64
-						if s := r.Schedules[id]; s != nil {
-							recipient.Recipient = s
-						} else {
-							recipientLogger.Warnw("ignoring unknown escalation recipient")
-							escalation.Recipients[i] = nil
-						}
-					} else {
-						recipientLogger.Warnw("ignoring unknown escalation recipient")
-						escalation.Recipients[i] = nil
-					}
-				}
+			elementRule.Escalations[newElement.ID] = newElement
+			return nil
+		},
+		func(curElement, update *rule.Escalation) error {
+			if curElement.RuleID != update.RuleID {
+				return removeAndAddInstead
+			}
 
-				escalation.Recipients = slices.DeleteFunc(escalation.Recipients, func(r *rule.EscalationRecipient) bool {
-					return r == nil
-				})
+			curElement.ChangedAt = update.ChangedAt
+			curElement.NameRaw = update.NameRaw
+			// Condition{,Expr} are being initialized by config.IncrementalConfigurableInitAndValidatable.
+			curElement.Condition = update.Condition
+			curElement.ConditionExpr = update.ConditionExpr
+			// TODO: synchronize Fallback{ForID,s} when implemented
+
+			return nil
+		},
+		func(delElement *rule.Escalation) error {
+			elementRule, ok := r.Rules[delElement.RuleID]
+			if !ok {
+				return fmt.Errorf("rule escalation refers unknown rule %d", delElement.RuleID)
 			}
 
-			if currentRule := r.Rules[id]; currentRule != nil {
-				*currentRule = *pendingRule
-			} else {
-				r.Rules[id] = pendingRule
+			delete(elementRule.Escalations, delElement.ID)
+			return nil
+		})
+
+	incrementalApplyPending(
+		r,
+		&r.ruleEscalationRecipients, &r.configChange.ruleEscalationRecipients,
+		func(newElement *rule.EscalationRecipient) error {
+			ok := false
+			if newElement.ContactID.Valid {
+				newElement.Recipient, ok = r.Contacts[newElement.ContactID.Int64]
+			} else if newElement.GroupID.Valid {
+				newElement.Recipient, ok = r.Groups[newElement.GroupID.Int64]
+			} else if newElement.ScheduleID.Valid {
+				newElement.Recipient, ok = r.Schedules[newElement.ScheduleID.Int64]
+			}
+			if !ok {
+				return fmt.Errorf("rule escalation recipient is missing or unknown")
 			}
-		}
-	}
 
-	r.pending.Rules = nil
+			newElement.Escalation = r.GetRuleEscalation(newElement.EscalationID)
+			if newElement.Escalation == nil {
+				return fmt.Errorf("rule escalation recipient refers to unknown escalation %d", newElement.EscalationID)
+			}
+			newElement.Escalation.Recipients = append(newElement.Escalation.Recipients, newElement)
+
+			return nil
+		},
+		nil,
+		func(delElement *rule.EscalationRecipient) error {
+			delElement.Escalation.Recipients = slices.DeleteFunc(
+				delElement.Escalation.Recipients,
+				func(recipient *rule.EscalationRecipient) bool {
+					return recipient.EscalationID == delElement.EscalationID
+				})
```

Comment:
```suggestion
			escalation := r.GetRuleEscalation(delElement.EscalationID)
			if escalation != nil {
				escalation.Recipients = slices.DeleteFunc(escalation.Recipients, func(recipient *rule.EscalationRecipient) bool {
					return recipient.EscalationID == delElement.EscalationID
				})
			}
```

There is absolutely no reason to introduce an escalation field to the recipients type.

Reply:
Totally valid point. The extra struct field seemed necessary in a way earlier state of this PR.